apply spacing rescaling when computing centroid_weighted

enricotagliavini opened the pull request:
## Description

We noticed the numbers for the _centroid_weighted_ returned by _regionprops_, when using anisotropic spacing (e.g. 2, 1 1), where not correct. Upon investigation it looks like the slice.start coordinates are not rescaled before adding to the, correctly rescaled, centroid_weighted_local. This commit fixes the issue and should now return the correct coordinates.

## For reviewers

<!-- Don't remove the checklist below. -->

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

enricotagliavini commented:
Yeah I noticed. I had a quick look but didn't figure it out yet. I'll try to have a look in the coming weeks, hopefully I can figure it out.

enricotagliavini commented:
Ok this passed the tests. It's a bit odd as the columns are the coordinate (e.g. (X Y Z)) for single channel images, but it becomes the channel index for multi channel images. I didn't know this when doing the first patch and it was tested only on single channel images. Although we always have multichannel images, we always analyze the channels separately. 

As for a test I can think of this: an image with only one non zero point in the corner furthest from the origin. With a spacing greater than 1 the centroid would end up outside the image. Would that be acceptable?

enricotagliavini commented:
Actually you already have a test for this, but the input image being used is not general enough to cover all cases. In particular for that image the slice.start is always the image origin, so it's 0. Even if not correctly multiplied by the spacing the end result is always correct as any number multiplied by zero yields 0.

lagru commented on skimage/measure/_regionprops.py:
```diff
@@ -631,8 +631,8 @@ def solidity(self):
         ctr = self.centroid_weighted_local
-        return tuple(idx + slc.start
-                     for idx, slc in zip(ctr, self.slice))
+        return tuple(idx + slc.start * spc
+                     for idx, slc, spc in zip(ctr, self.slice, self._spacing))
```
I'm probably not grasping all the bits and pieces properly yet but I'm confused by something. I checked how the spacing is taken into account in the non-weighted case in 
https://github.com/scikit-image/scikit-image/blob/2f9b82ce95e95c558ffe68bced6225c9a4560e31/skimage/measure/_regionprops.py#L433-L438

(called by `self.centroid`). There it looks like the spacing is multiplied with the sum of index and offset / `slice.start`, while in this fix it's only multiplied with `slice.start`.  Should it be 
```suggestion
        return tuple((idx + slc.start) * spc
                     for idx, slc, spc in zip(ctr, self.slice, self._spacing))
```
instead?

enricotagliavini replied:
The non-weighted centroid is computed in a much different way, as it's much simpler. It computes the centroid without taking into account spacing, and applies all the corrections at the very end.

When computing the weighted centroid we start from the local (to the prop bbox) centroid, which is already corrected for spacing, and we just need to apply a translation to the coordinate system of the original (full size) image. If we apply the correction again as suggested this would yield wrong coordinates, as the local centroid would be corrected twice.

More in details: the centroid weighted local uses the weighted moments

https://github.com/scikit-image/scikit-image/blob/2f9b82ce95e95c558ffe68bced6225c9a4560e31/skimage/measure/_regionprops.py#L638-L660

and the weighted moments just use the moments which take spacing into account (see line 261 in particular):

https://github.com/scikit-image/scikit-image/blob/2f9b82ce95e95c558ffe68bced6225c9a4560e31/skimage/measure/_moments.py#L154-L269

Does this clear the concern.

lagru replied:
Yes, I think I'm following now. Thanks for the detailed explanation and the contribution in total! :D

lagru reviewed (commented):
Thanks for working on this @enricotagliavini! If you merge the current `main` into this branch, the failing job on CircelCI should fix itself.

enricotagliavini commented:
> Thanks for working on this @enricotagliavini! If you merge the current `main` into this branch, the failing job on CircelCI should fix itself.

Thanks @lagru for checking this out and confirming this was a CI issue. I was quite puzzled for a bit. I addressed your comment on the fix, let me know if this clarifies.

stefanv commented:
Thank you for your contribution, @enricotagliavini.

enricotagliavini commented:
Thank you so much for accepting the fix. Cheers.
